include/crypto/event.h

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_INTERNAL_EVENT_H
+# define OSSL_INTERNAL_EVENT_H
+# pragma once
+
+# include <stddef.h>
+# include <stdarg.h>
+# include <openssl/safestack.h>
+# include "crypto/sparse_array.h"
+
+/*-
+ * Types and structures
+ * --------------------
+ */
+
+/*
+ * An event is expected to be very short lived, and may be a static
+ * (re-usable) variable, local in a function or other structure, or
+ * dynamically allocated.
+ */
+struct ossl_event_st;
+typedef int ossl_event_destructor_fn(struct ossl_event_st *event);
+
+/*
+ * Generic callback function type.  This is used for subscribers and filters
+ * alike, as seen in the different functions where this type is used.
+ */
+typedef int ossl_event_callback_fn(struct ossl_event_st *event,
```

Comment:
Perhaps document what the return means.  I see at least four options:
1. event processed and completed
2. event processed but not complete
3. event processed but an error occurred
4. I don't know about this event

The idea being that a completed event can short circuit the remaining of the processing and an event that was not know by any of the handlers could raise an assert (or error).  An error is likely a short circuit of the remaining handlers -- but we can discuss.

Reply:
That suggests that there would be an order between subscribers.  I was imagining it being done "in parallel", i.e. call each subscriber, and count errors and successes (where "pass", i.e. "I don't know about this event", is considered a success).

This doesn't, btw, answer what should actually happen when an error occur, or how that should be reported back.
One idea, which I admit stealing from [I-D.ietf-taps-interface](https://datatracker.ietf.org/doc/html/draft-ietf-taps-interface-15#appendix-A.2), is to have errors propagate as events.

Reply:
No order requires, just one subscriber needs to say it's all finished.

Reply:
> No order requires, just one subscriber needs to say it's all finished.

Which suggest that this particular subscriber needs to be called last, no?

Me, I'm rather thinking that when all subscribers have been called, this event is finished.

Reply:
Once it says that the event is dealt with, it will be last.
It if doesn't processing can continue.